### ISSUE TITLE
Heap corruption when passing vector of optional int64_t

### ISSUE THREAD

**Issue by Gugap** [jtv/libpqxx#906]
Running following:
```c++
pqxx::work transaction(connection);

std::vector<std::optional<int64_t>> test;
test.resize(300);
transaction.exec_params("INSERT INTO ...", test);
```
will crash due to writing into out of bounds - seems like the buffer size calculation is incorrect in this case.
For other types such as strings it works perfectly.

Happening on version 7.9.2.
Compiler: MSVC (most recent VS2022), lib integrated via vcpkg.

### UPDATE:

Can be reproduced in much simpler way:
```c++
std::vector<std::optional<int64_t>> test;
test.resize(10);
auto result = pqxx::to_string(test);
```

**Comment by tt4g** [comment]
pqxx is supposed to convert `std::vector` to binary data, but I think it is failing to do so: https://github.com/jtv/libpqxx/blob/7.9.2/include/pqxx/doc/binary-data.md
Can you check the error message with `try`-`catch`?

**Comment by Gugap** [comment]
I actually get an assert fired by std, due too heap corruption (out of bounds writing).
After further debugging, I found out the issue is caused at 
https://github.com/jtv/libpqxx/blob/6541c61dc245f3588f9e3d9888a606c5f5ddad13/include/pqxx/internal/conversions.hxx#L1236


more specifically: https://github.com/jtv/libpqxx/blob/6541c61dc245f3588f9e3d9888a606c5f5ddad13/include/pqxx/internal/conversions.hxx#L1102

where, if passing `std::vector<std::optional<int64_t>> test;` that has `nullopts`, where `std::size(s_null)` returns 3 bytes per `nullopt`.
Passing in vector of 10 `nullopts` will calculate buffer size of 33, but then the `string_traits<T>::into_buf(data, data + std::size(buf), value)};` will write 51 characters into the buffer : _{NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL}_

for `std::string`, the other std::accumulate branch is taken, which then calculates 10 bytes per `nullopt`, which is more than enough, so it seems like it is broken only for basic types that satisfy the `is_unquoted_safe`.

**Comment by tt4g** [comment]
The data in `std::vector<std::optional<int64_t>>` is assumed to be binary data, so `std::nullopt` may not be assumed.
What kind of data do you want to send to PostgreSQL with `std::vector<std::optional<int64_t>>`?

**Comment by Gugap** [comment]
Im doing batch insert of big ints  (`$1:bigint[]`), which sometimes can be NULL, which I use the `std::vector<std::optional<int64_t>>` for.

**Comment by Gugap** [comment]
I do not think that assumption is correct. 
`pqxx::to_string` actually has exactly same issue I described above (out of bounds write), as its the same function used to do the conversion, so its even an easier way to reproduce this issue. 
As the doc you  have linked (https://github.com/jtv/libpqxx/blob/7.9.2/include/pqxx/doc/binary-data.md) states, contiguous memory blocks of `uint8_t` or `std::byte` can be interpreted as bytes using `binary_cast`.

I have tried use `pqxx::to_string` on `std::vector<uint8_t>` and it is a disallowed conversion and fails to compile, which does make sense, to avoid incorrect use. Instead, you need to do `pqxx::to_string(pqxx::binary_cast(vecOfBytes))`

Moreover, you actually cannot `pqxx::binary_cast` `vector<int>` - there is a `static_assert `checking if the underlying type is size of 1!


As the tests you have also linked show, doing `pqxx::to_string()` on vector of ints should produce an array - i.e. **{5,4,3,2}**.
The same behavior happens for `std::optional<int>`, its only that when the `std::optional` is NULL, it incorrectly calculates size of the buffer (it still writes correct data into it, i.e. **{NULL,NULL,NULL,NULL,NULL}**! It's just too long and thats where the bug is).
`

**Comment by tt4g** [comment]
You are right. I had forgotten all about `pqxx::binary_cast`.

I checked the source code, when a value that maps to `NULL` is found in the container, `“NULL”` string is written to the buffer.

https://github.com/jtv/libpqxx/blob/6541c61dc245f3588f9e3d9888a606c5f5ddad13/include/pqxx/internal/conversions.hxx#L1047

The following `std::size(s_null)`  should return 4.

https://github.com/jtv/libpqxx/blob/6541c61dc245f3588f9e3d9888a606c5f5ddad13/include/pqxx/internal/conversions.hxx#L1102

Could you try to apply that patch and see if it solves the problem?

**Comment by tt4g** [comment]
This may be a bug in MSVC.

`s_null` is a subclass of `std::string_view` and its size is 4.

https://github.com/jtv/libpqxx/blob/6541c61dc245f3588f9e3d9888a606c5f5ddad13/include/pqxx/internal/conversions.hxx#L1020

The specification says that `std::size(const C& c)` must return `c.size()`: https://en.cppreference.com/w/cpp/iterator/size

```cpp
template< class C >
constexpr auto size( const C& c ) -> decltype(c.size());
```

Does your MSVC print `std::size(test)` 3 in the following code?

```cpp
#include <iterator>
#include <iostream>

int main(void)
{
    std::string_view test{"NULL"};
    std::cout << std::size(test) << std::endl;
    return 0;
}

**Comment by Gugap** [comment]
I think it does return 4 correctly, but:
https://github.com/jtv/libpqxx/blob/6541c61dc245f3588f9e3d9888a606c5f5ddad13/include/pqxx/internal/conversions.hxx#L1104
subtracts 1 from it, resulting in 3 bytes per `NULL`.

That being said, I feel like even when returning 4 it wouldnt be enough for the string, as there also need to be space for all the `,` between the `NULLs`. Potentially this meant to be +1 rather than -1? Then I think it would be the right amount. Unless im missing something.

**Comment by Gugap** [comment]
I also have a feeling that maybe the parentheses are incorrect and the -1 is only meant for the `elt_traits::size_buffer(elt)` to begin with?

**Comment by jtv** [comment]
Actually, isn't this much simpler?  String conversions were never meant to deal with null values.  A default-constructed `std::optional<...>` is a null value.

(And on top of that, a `std::vector<...>` for types other than `std::byte` translates to an SQL array.  SQL arrays can't contain nulls either.)

**Comment by jtv** [comment]
So perhaps the real problem is that there was _no error_ when you tried to convert a null value to a string!  There's a global check for nulls in `pqxx::to_string()` — but a lot of conversions now bypass that because they call `pqxx::string_traits<...>::to_buf()` or `pqxx::string_traits<...>::into_buf()`.

If that bears out, then I think what I need to do is...
1. Add a global `pqxx::into_buf()` (there already is a `pqxx::to_buf()`).
2. Make sure both throw an exception when asked to convert a null.
3. Always use use those global wrappers, not call the `string_traits<...>` implementations directly.

**Comment by jtv** [comment]
Unfortunately this also means that the "unnest" trick to insert a batch of data faster will not work in the presence of nulls.  SQL arrays simply don't support that.

What to do instead?  Consider using `stream_to`.  If that doesn't fit your use-case, e.g. because you need `ON CONFLICT UPDATE`, then perhaps use a `stream_to` to write into a temporary table first, and then a separate statement to merge that data into the ultimate destination table.

**Comment by Gugap** [comment]
The batch insert unnesting trick is exactly what im using this for:
```c++
std::vector<std::optional<std::string>> vecOfVals1;
std::vector<std::optional<std::string>> vecOfVals2;
std::vector<std::optional<int64_t>> vecOfVals3;
// .....
transaction.exec_params(
R"(with input_rows(val1, val2, val3) AS
(
    SELECT 
	 // .....
    FROM unnest(
        $1::text[], 
        $2::text[], 
        $3::bigint[], 
) as t(val1, val2, val3)        
),
 // followed by inserts

)", vecOfVals1, vecOfVals2, vecOfVals3);
```


As far as I can tell, it does work with NULLs perfectly (on the SQL side). On pqxx side, as long as there are NULLs in the text/string values (`std::vector<std::optional<std::string>>`) it works perfectly too. When NULLs appeared in the numeric vector, the issue described above appeared.

Now, I am by no means SQL expert, but to my understanding, the NULLs inside these array work just fine.
I also tried to verify with query as simple as: `select ARRAY[1, null, 2]::bigint[]` and it does seem to work. Im using PostgreSQL v16.4. Maybe its a feature added recently?


Regarding the NULL support not being intended in pqxx - sorry for my ignorance, but isnt this exactly kind of explained in the documentation?: https://github.com/jtv/libpqxx/blob/1ca80b0e638f6182426c5b11255069cae4fbd542/include/pqxx/doc/streams.md?plain=1#L25C40-L32C32
Now I certaintly dont want to come here and tell you whether the feature is intended in your library or not, but it would really be a shame if this (super handy and seemingly working, aside this specific case, which fails due to an incorrect size calculation) feature was discontinued!

I will also look into the `stream_to` solution, if thats something thats even more efficient way to do batch inserts, then great!
Though, at the end of the day, isnt it still needed to do the conversion from `optional`, `nullptr const char*`, `std::unqiue_ptr` or other supported nullable types to the actual 'NULL' string? Even with the stream API?

**Comment by jtv** [comment]
Ah you're right @Gugap — the ability for array elements to be null was added in 9.0.  (Which may not sound "recent" of course but libpqxx started well before that, and the old rule stuck in my memory.)

So it's kind of necessary to be able to convert null values to `NULL`.  But for strings, this should happen in cases where we support quoting of non-null values, since we want an SQL text of `NULL` and not `'NULL'`.

**Comment by jtv** [comment]
(I should probably explain that I designed the string conversion mechanism a long time ago, and it's grown fairly complex, so I often need to get re-read things when these questions come up.  Writing libpqxx is a labour of love and sadly I don't get to do it in my daily work.)

**Comment by jtv** [comment]
Ah yes.  There's a special provision _in the conversion from string to array_ to support null values.  The generic string conversion mechanism does not support converting a null to a string.

And as far as I can see now, the budget function for "unquoted-safe" types such as `std::optional<int64_t>` is just a little too tight.  The problem is that in adding up the budgets for the individual elements, I subtract the room for each element's terminating zero.  We don't actually want to include the elements' terminating zeroes.  But I should then add that one byte of budget right back, because we _do_ want _separators_ between the elements.

**Comment by jtv** [comment]
@Gugap perhaps you could try the change I'm proposing in #907?

**Comment by Gugap** [comment]
I am currently traveling and unfourtunately wont be able to test the fix in next couple of days - will do when back if still needed.

Just wanted to point out that during the testing I did: https://github.com/jtv/libpqxx/issues/906#issuecomment-2495993285

I noticed, that for vector of 10 nullopts, the `size_buffer` function was returning 33 (3 per element + 3 hardcoded at the start). 
The resulting string buffer that was produced by `string_traits<T>::into_buf(data, data + std::size(buf), value)};` was 51 characters long (    **{NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL}**   ). If Im reading the proposed change correctly, it removes the `-1`, which would result into 43 bytes for this scenario, I think, which would still be below the 51 needed (not accounting for the '`,`' between each element).
